### ISSUE TITLE
Failure in priority_update test

### ISSUE THREAD

**Issue by martinthomson** [mozilla/neqo#1304]
https://github.com/martinthomson/neqo/runs/4655353217?check_suite_focus=true#step:7:653 shows the following cryptic output:

```
---- priority_update stdout ----
thread 'priority_update' panicked at 'called `Option::unwrap()` on a `None` value', neqo-http3/tests/priority.rs:111:41
```

I've opened a PR for including a backtrace.  I can't see why this might have failed.  The test is pretty simple and I can't reproduce locally.

**Comment by KershawChang** [comment]
I somehow managed to reproduce this locally.
It seems the reason that we failed to get a [server event](https://searchfox.org/mozilla-central/rev/7271a078fa0c1b858a52614ea60ac82fdd8b3d23/third_party/rust/neqo-http3/tests/priority.rs#111) is related to [this line](https://searchfox.org/mozilla-central/rev/7271a078fa0c1b858a52614ea60ac82fdd8b3d23/third_party/rust/neqo-transport/src/recv_stream.rs#516).
I found that when this test failed, `already_data_ready` is true. So, there was [no stream readable event](https://searchfox.org/mozilla-central/rev/7271a078fa0c1b858a52614ea60ac82fdd8b3d23/third_party/rust/neqo-transport/src/recv_stream.rs#576) created. 

More details:
The `RecvStream` in successful case is:
```
RecvStream { stream_id: StreamId(2), state: Recv { fc: ReceiverFlowControl { subject: StreamId(2), max_active: 1048576, max_allowed: 1048576, consumed: 28, retired: 28, frame_pending: false }, session_fc: RefCell { value: ReceiverFlowControl { subject: (), max_active: 4611686018427387903, max_allowed: 4611686018427387903, consumed: 63, retired: 63, frame_pending: false } }, recv_buf: RxStreamOrderer { data_ranges: {}, retired: 28 } }, conn_events: ConnectionEvents { events: RefCell { value: [] } }, keep_alive: None }
```
The `RecvStream` in failed case is:
```
RecvStream { stream_id: StreamId(2), state: Recv { fc: ReceiverFlowControl { subject: StreamId(2), max_active: 1048576, max_allowed: 1048576, consumed: 26, retired: 23, frame_pending: false }, session_fc: RefCell { value: ReceiverFlowControl { subject: (), max_active: 4611686018427387903, max_allowed: 4611686018427387903, consumed: 61, retired: 58, frame_pending: false } }, recv_buf: RxStreamOrderer { data_ranges: {0: [0, 4, 11, 1, 64, 100, 7, 64, 100, 171, 96, 55, 66, 0, 204, 200, 238, 249, 46, 194, 22, 91, 0, 13, 1, 10]}, retired: 23 } }, conn_events: ConnectionEvents { events: RefCell { value: [] } }, keep_alive: None }
```

**Comment by martinthomson** [comment]
Hmm, that indicates that the server hasn't read everything out of the control stream yet.  Stream 2 is the first unidirectional stream the client opens, so I'm guessing that this is the control stream.  That's where PRIORITY_UPDATE is exchanged, so that's good.  Here's a breakdown of the stream data from the failed test:

0 - control stream
4 - SETTINGS
  11 - length
  1, 64, 100 - QPACK table size of 100
  7, 64, 100 - QPACK blocked streams at 100
  171, 96, 55, 66, 0, - disable WebTransport
204, 200, 238, 249, 46, 194, 22, 91, 0 - grease frame (using a maximum sized identifier, which is expected given how we construct these)
13, 1, 10 - MAX_PUSH_ID

I took a look at the code to work out maybe why the MAX_PUSH_ID frame wasn't being consumed, but that isn't it.  The problem isn't obvious, but the clue is that the grease frame is zero length.  We have a one in eight chance of that happening, which is why we don't see too many failures.

When that happens, our code for reading frames - specifically the part where we ignore unknown frame types - sets up to read zero bytes from the socket.  Zero is a special value there, so that does odd things to the code.

So I've a fix.

**Comment by ddragana** [comment]
Fixed by #1315
